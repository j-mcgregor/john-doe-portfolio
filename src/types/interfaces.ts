export interface NavbarProps {
    brand: string | JSX.Element
}

export interface PRISMIC_RichTextType {
    type: string
    text: string
    spans: any[]
}

export interface PRISMIC_Image {
    dimensions: {
        width: number
        height: number
    }
    alt?: string | null
    copyright: string | null
    url: string
}

export interface PRISMIC_ExperienceType {
    job_title: string
    date_from: string
    date_to?: string
    present?: boolean
    company: string
    city: string
    description: PRISMIC_RichTextType[]
}

export interface PRISMIC_landing_node {
    node: {
        primary_text: PRISMIC_RichTextType[]
        secondary_text: PRISMIC_RichTextType[]
        background_image: PRISMIC_Image
        about: PRISMIC_RichTextType[]
        experience: PRISMIC_ExperienceType[]
    }
}

export interface IndexPageProps {
    data: {
        prismic: {
            allLandings: {
                edges: PRISMIC_landing_node[]
            }
        }
        file: {
            childImageSharp: {
                fixed: {
                    height: number
                    width: number
                    base64: string
                    src: string
                    srcSet: string
                }
            }
        }
    }
}

export interface HeroProps {
    primary: JSX.Element
    secondary: JSX.Element
    logo?: JSX.Element
    backgroundImage?: string
}

export interface StyledHeroProps {
    bgImageUrl?: string
}

export interface AboutProps {
    about: JSX.Element
    experience: JSX.Element[]
}

// >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
// >>>>>>>>>>>>>>>>>/  FORMS   >>>>>>>>>>>>>>>>>>>>
// >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>

export interface TextInputProps {
    type: string
    placeholder: string
    label?: string
    name: string
    required: boolean
    classNames?: string[]
    value: string
    onChange: (e: React.FormEvent<HTMLInputElement>) => void
}

export interface TextAreaInputProps {
    placeholder: string
    label?: string
    name: string
    required: boolean
    value: string
    onChange: (e: React.FormEvent<HTMLTextAreaElement>) => void
}
