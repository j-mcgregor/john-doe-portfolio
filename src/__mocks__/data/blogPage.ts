import _ from 'lodash'
import {
    BlogPageProps,
    PRISMIC_PRIMARY_Image_with_Caption,
    PRISMIC_PRIMARY_Quote,
    PRISMIC_PRIMARY_Text,
} from '../../types/interfaces/blog'
import {
    PRISMIC_blog_node,
    PRISMIC_RichTextType,
    PRISMIC_MetaType,
} from '../../types/interfaces/prismic'

type ImageType = 'image_with_caption' | 'full-width-image' | 'emphasized'

type TextType = 'paragraph' | 'heading3'

export interface ImageSliceTypeProps {
    [index: string]: ImageType
}

export const imageSliceType: ImageSliceTypeProps = {
    '0': 'image_with_caption',
    '1': 'full-width-image',
    '2': 'emphasized',
}

export const makeTextType = (
    type: TextType,
    text: string
): PRISMIC_RichTextType => ({
    type,
    text,
    spans: [],
})

export const meta = (
    id: string,
    uid: string,
    type: string
): PRISMIC_MetaType => ({
    id,
    uid,
    type,
})

export const imageSlice = (
    imageType: number
): PRISMIC_PRIMARY_Image_with_Caption => ({
    primary: {
        caption: [makeTextType('paragraph', `caption-${imageType}`)],
        image: {
            dimensions: {
                width: 1200,
                height: 900,
            },
            alt: `alt-${imageType}`,
            copyright: null,
            url: 'http://fillmurray.com/300/300',
        },
    },
    label: `default`,
    type: imageSliceType[`${imageType}`],
})

export const quoteSlice = (text: string): PRISMIC_PRIMARY_Quote => ({
    primary: {
        quote: [makeTextType('paragraph', text)],
    },
    label: undefined,
    type: 'quote',
})

export const textSlice = (text: string): PRISMIC_PRIMARY_Text => ({
    primary: {
        text: [makeTextType('paragraph', text)],
    },
    label: undefined,
    type: 'text',
})

export const blogNodes: PRISMIC_blog_node[] = _.times(3, num => ({
    node: {
        date: `2020-0${num + 1}-01`,
        body: [
            textSlice('Lorem ipsum dolor sit amet'),
            imageSlice(num),
            quoteSlice('And I quoteth thus...'),
        ],
        _meta: meta(`id-${num + 1}`, `blog-${num + 1}`, 'blog_post'),
        title: [makeTextType('paragraph', `Post ${num + 1}`)],
    },
}))

export const blogPage: BlogPageProps = {
    data: {
        prismic: {
            allBlog_posts: {
                edges: blogNodes,
            },
        },
    },
}
