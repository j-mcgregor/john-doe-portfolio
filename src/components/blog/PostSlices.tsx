import React from 'react'
import {
    PRISMIC_PRIMARY_Image_with_Caption,
    PRISMIC_PRIMARY_Quote,
    PRISMIC_PRIMARY_Text,
} from '../../types/interfaces/blog'
import { ImageCaption, Quote, Text } from '../../components/shared/slices'
import createKey from '../../utils/createKey'

export interface PostSlicesProps {
    // INTERSECTION TYPES
    // https://www.typescriptlang.org/docs/handbook/advanced-types.html#intersection-types
    slices: Array<
        PRISMIC_PRIMARY_Image_with_Caption &
            PRISMIC_PRIMARY_Quote &
            PRISMIC_PRIMARY_Text
    >
}

// Sort and display the different slice options
const PostSlices: React.FC<PostSlicesProps> = ({ slices }) => {
    const postSlices = slices.map((slice, index) => {
        switch (slice.type) {
            case 'text':
                return (
                    <div
                        key={createKey(slice.type, index)}
                        className="homepage-slice-wrapper"
                    >
                        <Text slice={slice} />
                    </div>
                )

            case 'quote':
                return (
                    <div
                        key={createKey(slice.type, index)}
                        className="homepage-slice-wrapper"
                    >
                        <Quote slice={slice} />
                    </div>
                )

            case 'image_with_caption':
                return (
                    <div
                        key={createKey(slice.type, index)}
                        className="homepage-slice-wrapper"
                    >
                        <ImageCaption slice={slice} />
                    </div>
                )

            default:
                return <div key={createKey(slice.type, index)} />
        }
    })

    return <div>{postSlices}</div>
}

export default PostSlices
