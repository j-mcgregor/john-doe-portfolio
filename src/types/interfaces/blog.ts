// >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
// >>>>>>>>>>>>>>>>>>>  BLOG   >>>>>>>>>>>>>>>>>>>>
// >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>

import {
    PRISMIC_RichTextType,
    PRISMIC_Image,
    PRISMIC_blog_node,
    PRISMIC_blog_node_body,
} from './prismic'

export interface BlogBodyProps {
    __typename?: string
    type: string
    primary: {
        text?: PRISMIC_RichTextType[]
        caption?: string | null
        image?: PRISMIC_Image
    }
}

export interface BlogPageProps {
    data: {
        prismic: {
            allBlog_posts: {
                edges: PRISMIC_blog_node[]
            }
        }
    }
}

export interface PostSummaryProps {
    post: PRISMIC_blog_node_body
}

export interface BlogPostsProps {
    posts?: PRISMIC_blog_node[]
}
