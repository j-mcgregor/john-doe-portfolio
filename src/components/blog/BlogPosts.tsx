import React from 'react'
import { BlogPostsProps } from '../../types/interfaces/blog'
import { PostItem } from './PostItem'

const BlogPosts: React.FC<BlogPostsProps> = ({ posts }) => {
    if (!posts) return null

    return (
        <div className="blog pb7">
            <h1>Blog</h1>
            {posts.map(post => (
                <PostItem post={post.node} key={post.node._meta.id} />
            ))}
        </div>
    )
}

export default BlogPosts
