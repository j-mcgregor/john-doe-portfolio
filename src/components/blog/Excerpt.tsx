import React from 'react'
import { BlogBodyProps, PostSummaryProps } from '../../types/interfaces/blog'
import { RichText } from 'prismic-reactjs'

// Function to retrieve a small preview of the post's text
export const Excerpt: React.FC<PostSummaryProps> = ({ post }) => {
    // Find the first text slice of post's body
    const firstTextSlice = post.body.find(
        (slice: BlogBodyProps) => slice.type === 'text'
    )

    if (firstTextSlice != null) {
        // Set the character limit for the text we'll show in the homepage
        const textLimit = 300
        const text = RichText.asText(firstTextSlice.primary.text)
        const limitedText = text.substring(0, textLimit)

        if (text.length > textLimit) {
            // Cut only up to the last word and attach '...' for readability
            return (
                <p>
                    {`${limitedText.substring(
                        0,
                        limitedText.lastIndexOf(' ')
                    )}...`}
                </p>
            )
        } else {
            // If it's shorter than the limit, just show it normally
            return <p>{text}</p>
        }
    } else {
        // If there are no slices of type 'text', return nothing
        return null
    }
}
