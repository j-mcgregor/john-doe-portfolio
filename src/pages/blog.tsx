import React from 'react'
import { graphql } from 'gatsby'

import Layout from '../components/Layout'
import SEO from '../components/SEO'
import { BlogPageProps } from '../types/interfaces/blog'
import BlogPosts from '../components/blog/BlogPosts'

export const query = graphql`
    query BlogPostsQuery {
        prismic {
            allBlog_posts {
                edges {
                    node {
                        title
                        date
                        body {
                            ... on PRISMIC_Blog_postBodyImage_with_caption {
                                primary {
                                    caption
                                    image
                                }
                                type
                            }
                            ... on PRISMIC_Blog_postBodyText {
                                primary {
                                    text
                                }
                                type
                            }
                        }
                        _meta {
                            id
                            type
                            uid
                        }
                    }
                }
            }
        }
    }
`

const BlogPage: React.FC<BlogPageProps> = ({
    data: {
        prismic: {
            allBlog_posts: { edges },
        },
    },
}) => {
    return (
        <Layout>
            <SEO title="Blog Post" />
            <BlogPosts posts={edges} />
        </Layout>
    )
}

export default BlogPage
