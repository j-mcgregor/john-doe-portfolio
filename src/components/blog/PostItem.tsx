import React from 'react'
import { Link } from 'gatsby'
import { Excerpt } from './Excerpt'
import { RichText, Date } from 'prismic-reactjs'
import { linkResolver } from '../../utils/RichTextCustom/linkResolver'
import { PostSummaryProps } from '../../types/interfaces/blog'

// A summary of the Blog Post
export const PostItem: React.FC<PostSummaryProps> = ({ post }) => {
    // Store and format the blog post's publication date
    let postDate = Date(post.date)

    postDate = postDate
        ? new Intl.DateTimeFormat('en-US', {
              month: 'short',
              day: '2-digit',
              year: 'numeric',
          }).format(postDate)
        : ''

    return (
        <div className="post-summary" key={post._meta.id}>
            <h2>
                {/* We render a link to a particular post using the linkResolver for the url and its title */}
                <Link to={linkResolver(post._meta)}>
                    {RichText.asText(post.title).length !== 0
                        ? RichText.asText(post.title)
                        : 'Untitled'}
                </Link>
            </h2>
            <p className="blog-post-meta">
                <time>{postDate}</time>
            </p>
            {/* Renders a small preview of the post's text */}
            <Excerpt post={post} />
        </div>
    )
}
