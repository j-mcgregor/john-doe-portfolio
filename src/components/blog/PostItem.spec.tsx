import React from 'react'
import { render } from '@testing-library/react'
import { PostItem } from './PostItem'
import { PostSummaryProps } from '../../types/interfaces/blog'
import { blogPage } from '../../__mocks__/data/blogPage'

describe('PostItem', () => {
    let props: PostSummaryProps

    beforeEach(() => {
        props = {
            post: {
                ...blogPage.data.prismic.allBlog_posts.edges[0].node,
            },
        }
    })

    it('should render a post item', () => {
        const { baseElement } = render(<PostItem {...props} />)

        const date = baseElement.getElementsByTagName('time')[0]
        expect(date.innerHTML).toBe('Jan 01, 2020')
        expect(baseElement).toMatchSnapshot()
    })

    it('should show blank for date if none provided', () => {
        const newProps = {
            post: {
                ...blogPage.data.prismic.allBlog_posts.edges[0].node,
                date: '',
            },
        }

        const { baseElement } = render(<PostItem {...newProps} />)

        expect(
            baseElement.getElementsByTagName('time')[0].innerText
        ).toBeUndefined()
    })

    it('should show Untitled for title if none provided', () => {
        const newProps = {
            post: {
                ...blogPage.data.prismic.allBlog_posts.edges[0].node,
                title: [],
            },
        }

        const { queryByText } = render(<PostItem {...newProps} />)

        const title = queryByText('Undefined')

        expect(title).toBeDefined()
    })
})
