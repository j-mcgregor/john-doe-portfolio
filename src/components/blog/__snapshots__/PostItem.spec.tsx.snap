// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostItem should render a post item 1`] = `
<body>
  <div>
    <div
      class="post-summary"
    >
      <h2>
        <a
          href="/blog/blog-1"
        >
          First post
        </a>
      </h2>
      <p
        class="blog-post-meta"
      >
        <time>
          Jan 01, 2020
        </time>
      </p>
      <p>
        Lorem ipsum dolor sit amet
      </p>
    </div>
  </div>
</body>
`;
